### PR TITLE
Special repository for ST clients

### PR DESCRIPTION
**Pull request by im-konge**
Signed-off-by: Lukas Kral <lukywill16@gmail.com>

This proposal suggests creating a new repository for `Strimzi` ST client based on 
[Strimzi client-examples](https://github.com/strimzi/client-examples).

### CODE REVIEW

**Review by scholzj — COMMENTED** [review]
I think you should clarify what does _based on Strimzi client-examples_ really mean. I assume what you are trying to say is that you will start by copying the existing example clients, but then they will each go their own way?

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (Frawless) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:49]: So if we will do Strimzi 0.22.2 and master will use Kafka 2.8, `systemtest-client` as well, how you will use `latest version supported by Strimzi` without additional versioning for `systemtest-client` ?

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (Frawless) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:9]: I think you could describe briefly `test-client` and why it's not sufficient/you want to replace it

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (Frawless) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:23]: Maybe you could put there more details? Like repo name, etc. Maybe also suggest component owners?

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (tombentley) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:32]:
```suggestion
    * component owners will be same as for STs
```

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (tombentley) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:35]:
```suggestion
       * build - `mvn` build, checkstyle (with some simple UTs or ITs)
```

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (tombentley) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:58]: If it's coupled tightly to the Strimzi operators ST then it should be in that repo and thus automatically always in sync. If you want it in a separate repo for use with the bridge, or other repos, then you should aim to have a versioned API (of some sort) and it's own release cadence (which you can chose based on new features, of course).

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (tombentley) [reply 021-special-repository-for-st-clients-based-on-example-clients.md:62]: I read this to mean it will always support a Kafka version that's at least as recent as that supported by the operator. There's no reason why it can't be more recent. And, there's actually some value in being able to test old clients against the new broker (which you could do by having multiple jars in the image and selecting them according to an env var).

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (scholzj) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:31]: Maybe we could just name it `test-clients`?

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (scholzj) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:58]: How would you implement this?

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (scholzj) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:61]: main branch now uses only 2.6.0+.

Comment (im-konge) [reply 021-special-repository-for-st-clients-based-on-example-clients.md:61]: This is meant for latest released Strimzi - `0.22.1` - where the `2.5.0` is supported, maybe I should specify it.

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (scholzj) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:60]:
```suggestion
For each supported Kafka version will be separate tag - pattern will be `test-client:TEST_CLIENTS_VERSION-kafka-KAFKA_VERSION`.
```

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (scholzj) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:59]: Not sure it is that easy ... maybe you should remove this and leave it more open to how it will be done or add more details.

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (Frawless) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:21]:
```suggestion
As we discussed offline, the `client-examples` repository should be really _exemplary_, 
```

**Review by Frawless — APPROVED** [review]
Just one nit. Otherwise, LGTM and I agree with that approach.

**Review by sknot-rh — APPROVED** [review]

**Review by scholzj — APPROVED** [review]

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (tombentley) [inline 021-special-repository-for-st-clients-based-on-example-clients.md:21]: I don't like to see mention of offline discussion: It suggests things are happening in private, for unknown reasons. I understand that sometimes offline discussion just happens in a non-public forum by accident. In those cases the discussion and outcome should be summarised publicly somewhere for anyone else to contribute. 

In this case I've no idea who "we" means, and in any case I don't think it matters: It is enough to say that client-examples should be exemplary, and that a dedicated test client is better than pushing the example clients in a direction that makes them less useful as examples.

**Review by tombentley — APPROVED** [review]
SGTM, but please revise the "discussed offline" part.

File: 021-special-repository-for-st-clients-based-on-example-clients.md
Comment (im-konge) [reply 021-special-repository-for-st-clients-based-on-example-clients.md:21]: Good point, thanks! :)

**Comment by im-konge** [comment]
@ppatierno @samuel-hawker when you'll have time, can you take a look at this please?

**Review by ppatierno — APPROVED** [review]

**Comment by scholzj** [comment]
@im-konge It looks like we have enough approvals for this. So we can maybe give @samuel-hawker till the end of the day. But can you please update the index in README.md to add this proposal to it as well?

**Comment by im-konge** [comment]
@scholzj thanks, I updated the `README.md`. I agree with you on waiting till the end of the day for @samuel-hawker to review this. Thanks :)

**Pull request merged**
